From 2166055b0a30bdfc033769e1001e9852676d14df Mon Sep 17 00:00:00 2001
From: Vitaly Prosyak <vitaly.prosyak@amd.com>
Date: Tue, 11 Oct 2016 13:24:01 -0500
Subject: [PATCH 1223/1722] drm/amd/dal: Fix null stream on disable output

Add streams after required streams from new and replaced targets
are removed from freesync module.

Change-Id: I900e107d29918fefaebd93085f0037b75649a303
Signed-off-by: Vitaly Prosyak <vitaly.prosyak@amd.com>
Acked-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c | 21 +++++++++++++++------
 1 file changed, 15 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
index 7314412..3019c00 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
@@ -2384,6 +2384,7 @@ int amdgpu_dm_atomic_commit(
 
 	struct dc_target *commit_targets[MAX_TARGETS];
 	struct amdgpu_crtc *new_crtcs[MAX_TARGETS];
+	struct dc_target *new_target;
 
 	/* In this step all new fb would be pinned */
 
@@ -2444,7 +2445,7 @@ int amdgpu_dm_atomic_commit(
 		case DM_COMMIT_ACTION_DPMS_ON:
 		case DM_COMMIT_ACTION_SET: {
 			struct dm_connector_state *dm_state = NULL;
-			struct dc_target *new_target = NULL;
+			new_target = NULL;
 
 			if (aconnector)
 				dm_state = to_dm_connector_state(aconnector->base.state);
@@ -2492,11 +2493,6 @@ int amdgpu_dm_atomic_commit(
 			acrtc->enabled = true;
 			acrtc->hw_mode = crtc->state->mode;
 			crtc->hwmode = crtc->state->mode;
-			if (adev->dm.freesync_module)
-				for (j = 0; j < acrtc->target->stream_count; j++)
-					mod_freesync_add_stream(
-							adev->dm.freesync_module,
-							acrtc->target->streams[j]);
 			break;
 		}
 
@@ -2535,6 +2531,19 @@ int amdgpu_dm_atomic_commit(
 			++commit_targets_count;
 		}
 	}
+	/*
+	 * Add streams after required streams from new and replaced targets
+	 * are removed from freesync module
+	 */
+	if (adev->dm.freesync_module) {
+		for (i = 0; i < new_crtcs_count; i++) {
+			new_target = new_crtcs[i]->target;
+			for (j = 0; j < new_target->stream_count; j++)
+				mod_freesync_add_stream(
+						adev->dm.freesync_module,
+						new_target->streams[j]);
+		}
+	}
 
 	/* DC is optimized not to do anything if 'targets' didn't change. */
 	dc_commit_targets(dm->dc, commit_targets, commit_targets_count);
-- 
2.7.4

